content/blog/2025-dont-rerender-new-flatlist-items/index.mdx

```diff
@@ -0,0 +1,164 @@
+image: header.jpg
+author: Szymon Koper
+---
+
+#### Introduction
+
+This post is a follow-up to **[Don't Re-Render All FlatList Items](/blog/2022-dont-rerender-all-flatlist-items)**. In the previous article, we optimized FlatList to prevent unnecessary re-renders when **updating list item values**. This time, we’ll tackle another common issue: **avoiding re-renders when adding or removing items**.
+
+Many developers believe they need to **wrap every FlatList prop in `useCallback` or `useMemo`** to optimize performance. But is that really necessary? Let’s find out.
+
+This issue is especially common in **apps with infinite scrolling or pagination**, where new items are fetched and added to a list dynamically. If your app uses `onEndReached` in `FlatList` to fetch new pages from an API, the same principles apply.
+
+---
+
+### **The Problem: Every Item Re-Renders When Adding a New One**
+
+Imagine an app that fetches data from the internet and **appends new items** to a FlatList. Here’s what should happen:
+
+✅ **New items should render** when added.
+❌ **Existing items should not re-render** (since they haven’t changed).
+
+But in many cases, **all items re-render when a new one is added!**
+
+#### **Let's See It in Action**
+
+<div class="gif-container">
+
+![Recording of the example app - adding and deleting items](flatlist-app-demo.gif)
+
+This is the example app where items are added and removed.
+
+</div>
+
+Here’s an Expo Snack where **all items re-render** when we add a new one:
+
+👉 **[Try the Snack Here](https://snack.expo.dev/@sakydpozrux/rerender-only-changed-items)** (You'll edit it yourself later!)
+
+**How to test:**
+1. Open the **console logs** in your browser.
+2. Click "Add Item" and watch the logs.
+3. Notice that **every existing item re-renders, even though they haven’t changed!**
+
+<div class="gif-container">
+
+![Logs before optimization - too many re-renders](flatlist-logs-before.gif)
+
+Here’s what happens in the console when adding items in the unoptimized version.
+
+</div>
+
+**Example console output:**
+```
+Rendering item: 1
+Rendering item: 2
+Rendering item: 1  <- Why is this rendering again?
+Rendering item: 2
+Rendering item: 3  <- Only this should be new
+Rendering item: 1
+Rendering item: 2
+Rendering item: 3
+```
+
+---
+
+### **The Fix: Just Add `React.memo`**
+
+Here’s the best part: the **fix is incredibly simple**.
+
+**✅ Wrap `ListItem` in `React.memo`. That’s it.**
+
+Now, `ListItem` will **only re-render when its props change**. Since each item’s `id` remains the same, old items won’t re-render when new ones are added.
+
+#### **Console Output After the Fix:**
+```
+Rendering item: 1
+Rendering item: 2
+Rendering item: 3  <- Only new items render now!
+```
+
+<div class="gif-container">
+
+![Logs after optimization - only new items render](flatlist-logs-after.gif)
+
+After the fix, only new items render as expected.
+
+</div>
+
+**Boom! No more unnecessary re-renders. 🎉**
+
```

Comment:
Technically, everything still re-renders:

* renderItem still runs for each list element
* just the component comes from cache